common/src/main/java/io/netty/util/internal/InternalThreadLocalMap.java

```diff
@@ -164,7 +165,7 @@ public int size() {
 
     public StringBuilder stringBuilder() {
         StringBuilder builder = stringBuilder;
-        if (builder == null) {
+        if (builder == null || builder.capacity() > STRING_BUILDER_MAX_CAPACITY /* ensure memory overhead */ ) {
```

Comment:
I don't think you want to create a new `StringBuilder` but instead trim it to some size if it exceeds that size.